<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Bwbble by viq854</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>Bwbble</h1>
        <p></p>

        <p class="view"><a href="https://github.com/viq854/bwbble">View the Project on GitHub <small>viq854/bwbble</small></a></p>


        <ul>
          <li><a href="https://github.com/viq854/bwbble/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/viq854/bwbble/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/viq854/bwbble">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h3>BWBBLE: Short Read Alignment with Populations of Genomes</h3>

<p>

<h3>Description</h3>

<p>
BWBBLE is a BWT-based aligner that maps short reads to a compressed linear representation of a collection of genomes
(the reference 'multi-genome').  
</p>

<p>
The reference multi-genome is created from a single genome build (e.g. GRCh37) augmented with a set of genomic variants 
obtained from a given collection of genomes (e.g. the 1000 Genomes Project). The SNPs present in the population 
are incorporated by extending the 4-letter A/C/T/G nucleotide code of the single reference to the 16-letter IUPAC 
code, which allows us to encode which nucleotides have been observed at a given position across all the available 
sequenced genomes (for example, the IUPAC character R can be used to represent that both A and G were observed). 
Indels are incorporated into the multi-genome by appending them at the end of the reference padded at both ends 
with the bases surrounding the indel at its position in the reference.
</p>

<p>
The reference multi-genome is stored using the standard FASTA file format. We provide a script to generate this FASTA file 
from a single reference FASTA file (e.g. GRCh37) and a set of VCF files containing the SNPs and indels to be 
incorporated into the reference multi-genome.
</p>

<p>
Given the reference multi-genome, the BWBBLE aligner operates in 3 steps:  
<br>
1. reference indexing (this creates the BWT of the reference multi-genome and auxiliary data structures to enable read mapping)
<br>
2. read mapping
<br>
3. alignment results evaluation and reporting
</p>

<p>
The reads are accepted in the FASTQ file format and the alignment results are reported using the SAM file format.
</p>

<p>
The BWBBLE read alignment alrogithm is based on the BWT backwards search procedure, with the difference that a read 
nucleotide can now match multiple bases in the reference multi-genome (namely, all the IUPAC codes that include this 
nucleotide; e.g. read base A can match the IUPAC characters A, D, H, M, N, R, V and W).
In order to tolerate sequencing errors and other variations of the reads from the reference multi-genome, 
similar to other aligners, BWBBLE allow up to a certain number of mismatches and gaps. 
Our inexact read alignment to the multi-genome reference algorithm is an extension of the inexact search algorithm 
used by BWA (Li and Durbin, 2009) and is guaranteed to find all the alignments with up to n differences. 
</p>

<p>
In order to reduce the search space and improving the performance of the alignment algorithm we have also adapted 
many heuristics from the existing BWA aligner. In particular, no sub-optimal alignments are explored 
if the number of best hits exceeds a given threshold. Furthermore, if the number of differences in the best hit, n_best, 
is less than n, then n is reset to n_best + 1. It is also possible to limit how many differences are allowed in 
the seed sequence (the first k base pairs at the beginning of the read, where k is
the seed length and can be adjusted) and disallow insertions and deletions at the ends of the read. 
Other search parameters available to the users of BWA (e.g. mismatch, gap open and gap extension penalties) 
have also been incorporated to provide a similar user interface and improve the efficiency of the BWBBLE implementation.

</p>

<!-- <pre><code>$ cd your_repo_root/repo_name
$ git fetch origin
$ git checkout gh-pages
</code></pre> -->

<!-- <p>This website is currently under construction.</p> -->

<h3>Command Line Arguments </h3>

Usage:   bwbble command [options]
<br>
<br>
Command: <br>  
&emsp; <b>index </b> &emsp; index the reference multi-genome in the FASTA format <br>
&emsp; bwbble index <i>seq_fasta</i> <br> <br>
	        
&emsp; <b>align</b> &emsp; align reads <br>
&emsp; bwbble align [options] <i>seq_fasta reads_fastq</i> [<i>output_aln</i>] <br>
&emsp;<b>Options</b>: <br> 
&emsp;&emsp; <b>M</b>&emsp;mismatch penalty <br>
&emsp;&emsp; <b>O</b>&emsp; gap open penalty <br>
&emsp;&emsp; <b>E</b>&emsp; gap extend penalty <br>
&emsp;&emsp; <b>n</b>&emsp; maximum number of differences in the alignment (gaps and mismatches) <br>
&emsp;&emsp; <b>l</b>&emsp;  length of the seed (seed := first l chars of the read) <br>
&emsp;&emsp; <b>k</b>&emsp; maximum number of differences in the seed <br>
&emsp;&emsp; <b>o</b>&emsp; maximum number of gap opens <br>
&emsp;&emsp; <b>e</b>&emsp; maximum number of gap extends <br>
&emsp;&emsp; <b>t</b>&emsp; run multi-threaded with t threads <br>
&emsp;&emsp; <b>S</b>&emsp;align with a single-genome reference <br><br>
          
          
&emsp; <b>aln2sam</b>  convert alignment results to SAM file format for single-end mapping<br>
&emsp; bwbble aln2sam [options] <i>seq_fasta reads_fastq alns_aln</i>

<h3>Examples</h3>

<h3>Authors and Contributors</h3>

<h3>Support or Contact</h3>

      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/viq854">viq854</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>
